site/journal/2020/Understanding args in Glimmer Components.md

```diff
@@ -0,0 +1,323 @@
+    this.#args = args;
+  }
+}
+```
+
+With that in place, only the `Component` base class can change the value of `args`: private class fields are not accessible from child classes. This is all we need to see what patterns do and don’t work with `args` in Glimmer components!
+
+Let’s imagine that we have some bucket of root state representing a user, with a name and an age. In a real app, this might correspond to data that comes from a [`Route` model hook](https://guides.emberjs.com/v3.22.0/routing/specifying-a-routes-model/), or which is stored on a service, or some other means. For our purposes, though, a simple class with some properties and methods to update them will do:
+
+```js
+class Root extends Component {
+  user = {
+    name: 'Chris Krycho';
+    age: 33;
+  };
+
+  haveABirthday() {
+    this.user.age++;
+  }
+
+  changeNameTo(newName) {
+    this.user.name = newName;
+  }
+}
+```
+
+Now let’s create a `Profile` component class which we can use to manage displaying a user’s profile info.  Here we are just extending the component definition we supplied above.
+
+```js
+class Profile extends Component {
+  constructor(args) {
+    super(args);
+  }
+
+  get description() {
+    return `${this.args.name} is ${this.args.age} years old`;
+  }
+}
+```
+
+Now we can create an instance of the `Root` and an instance of `Profile`, and pass the `user` in directly as the value for `args`. (There is a subtlety here in how this works in real Glimmer components when passing plain values rather than objects, which I cover in Part II below.)
+
+```js
+let root = new Root();
+let profile = new Profile(root.user);
+console.log(profile.description);
+// -> "Chris Krycho is 33 years old"
+```
+
+Now if we change the values of the properties on `root` by calling its methods, when we ask for `description` again it will “automatically” have the right value:
+
+```js
+root.haveABirthday();
+root.changeNameTo("C. D. Krycho");
+console.log(profile.description);
+// -> "C. D. Krycho Krycho is 34 years old"
+```
+
+But the “automatically” here is just normal JavaScript semantics: we passed in a reference to an object, so when we change the properties on that object and then invoke a getter which depends on those values, we get an updated value. While there are some differences in the specific details of the implementation (as I cover below in Part II), this is fundamentally how `args` works in Glimmer components. And notice: there’s no `@tracked` in sight, no template layer involved, and in fact nothing about Ember or Glimmer in sight. It is, as they say, *just JavaScript*™.
+
+This also means that we can see why certain things *won’t* work. I often see people write code like this both when writing new Glimmer components from scratch and when migrating from Ember components:
+
+```js
+class Profile extends Component {
+  constructor(args) {
+    super(args);
+    this.description =
+      `${this.args.name} is ${this.args.age} years old`;
+  }
```

Comment:
I know you explain this below why doing this doesn't actually pick up changes from `name` and `age`. However, I think we can make this even more obvious by re-references how example Part 1a worked (e.g. with a `getter`). Here's an example of how I would make it more obvious, but feel free to adapt it based on how you see fit:

```
class Profile extends Component {
  constructor(args) {
    super(args);
    this.description =
      `${this.args.name} is ${this.args.age} years old`; // Doesn't pick up changes to name and age
  }

  // Below actually picks up changes:
  // get description() {
  //   return `${this.args.name} is ${this.args.age} years old`;
  // }
}
```
To explain my own reading psychology, by the time I get to this section, I store very little in my brain of how Part 1a's example looks (e.g. with getter), so when it comes time to do a "diff" between Part 1a and 1b (e.g. assigning to an instance property in the constructor), I'd have to scroll up and find what 1a looks like. By having this in 1 place, it'll save me some trouble digging through the post, making it easier for lazy readers like me.

Reply:
Incorporated in a6065a6 – this was helpful!